frontend-react/src/components/Admin/AdminReceiverDashboard.tsx

```diff
@@ -0,0 +1,775 @@
+        let rowReceiver = currentReceiver; // used to know when we've run out of row data
+
+        // loop over all days
+        const daySlots = new TimeSlots([startDay, endDay], 24);
+        for (let [daySlotStart, daySlotEnd] of daySlots) {
+            const successForDay = new SuccessRateTracker();
+            const timeSlots = new TimeSlots(
+                [daySlotStart, daySlotEnd],
+                SKIP_HOURS
+            );
+            for (let [timeSlotStart, timeSlotEnd] of timeSlots) {
+                const successForSlice = new SuccessRateTracker();
+
+                // loop over keys that are in this range. Build aggregates
+                while (
+                    keyOffset < keys.length &&
+                    dateIsInRange(currentDate, [timeSlotStart, timeSlotEnd]) &&
+                    rowReceiver === currentReceiver
+                ) {
+                    // because we're moving two cursors through data (date slots and data results),
+                    // we use the data results from the prior iteration when we moved that cursor forward.
+                    const wasSuccessful =
+                        currentEntry.connectionCheckSuccessful;
+                    successForSlice.updateState(wasSuccessful);
+                    successForDay.updateState(wasSuccessful);
+                    successForRow.updateState(wasSuccessful);
+
+                    // next entry
+                    keyOffset++;
+                    if (keyOffset >= keys.length) {
+                        break;
+                    }
+                    currentKey = keys[keyOffset];
+                    currentEntry = props.data[currentKey];
+                    currentDate = new Date(
+                        currentEntry.connectionCheckCompletedAt
+                    );
+                    currentReceiver = `${currentEntry.organizationName}|${currentEntry.receiverName}`;
+                }
+
+                // render slots for day
+                {
+                    const sliceClassName =
+                        SUCCESS_RATE_CLASSNAME_MAP[
+                            successForSlice.currentState
+                        ];
+
+                    // we can use the currentkey in the data dict as a unique key for this cell
+                    sliceElements.push(
+                        <Grid
+                            row
+                            key={`slice:${currentReceiver}|${timeSlotStart}`}
+                            className={`slice ${sliceClassName}`}
+                            data-keyoffset={keyOffset - 1}
+                            onClick={(evt) => {
+                                if (
+                                    successForSlice.currentState ===
+                                    SuccessRate.UNDEFINED
+                                ) {
+                                    return;
+                                }
+                                const clickKey =
```

Comment:
[another nit] `parseInt("") === NaN; keys[NaN] === undefined` so, why not
```suggestion
                                const clickKey = evt.currentTarget?.dataset["keyoffset"] ? parseInt(evt.currentTarget.dataset["keyoffset"]) : "";
```
this is a dumb comment, but presenting an explicit code path that results in evaluating `parseInt("")` and then using that to index into an array makes me itchy

Reply:
Good feedback. Done